adding Java code (#721)
corda/corda-docs-portal: content/en/platform/corda/5.0-dev-preview-2/getting-started/fast-feedback-with-the-simulator/java-flow-test-code.md

---
date: '2022-09-20'
title: "Java Flow Test Code"
menu:
  corda-5-dev-preview2:
    parent: corda-5-dev-preview-simulator
    identifier: corda-5-dev-preview-flow-test-java
    weight: 2000
section_menu: corda-5-dev-preview2
---

> First the test class instantiates `MemberX500Name` for two actors. `MemberX500Name` is the primary way that identities are represented on a Corda [application Network](../../introduction/key-concepts.html#application-network). `MemberX500Name` has a static method `parse()` that turns the string representation of a members identity into a `MemberX500Name` object.

Reviewer: ```suggestion
First the test class instantiates `MemberX500Name` for two actors. `MemberX500Name` is the primary way that identities are represented on a Corda [application network](../../introduction/key-concepts.html#application-network). `MemberX500Name` has a static method `parse()` that turns the string representation of a members identity into a `MemberX500Name` object.
```

1. Set up Alice and Bob identities:
   ```java
   class MyFirstFlowTest {

>        // Names picked to match the corda network in config/dev-net.json

Reviewer: ```suggestion
       // Names picked to match the Corda network in config/dev-net.json
```

       private MemberX500Name bobX500 = MemberX500Name.parse("CN=Bob, OU=Test Dept, O=R3, L=London, C=GB");
   ...
   }
   ```
2. Declare the test method in the standard way for JUnit test, with a `@Test` annotation:
   ```java
   @Test
   @SuppressWarnings("unchecked")
   public void test_that_MyFirstFLow_returns_correct_message() {
     ...
   }
   ```      
3. Instantiate a `Simulator` class:
   ```java
   // Instantiate an instance of the Simulator
   Simulator simulator = new Simulator();
    ```
   To find out more about this class, read the [README.md from the source code](https://github.com/corda/corda-runtime-os/blob/release/os/5.0/simulator/README.md).

4. Convert the `MemberX500Name` for Alice and Bob into `HoldingIdentities` using the `HoldingIdentity` static method. The `HoldingIdentity` in Corda holds the `MemberX500Name` and the `GroupId` which is the unique identifier for the application network. However, the class used here is the simulated version provided by Simulator; that is `net.corda.simulator.HoldingIdentity` rather than `net.corda.virtualnode.HoldingEntity`.
   ```java

>    // Create Alice's and Bob HoldingIDs

Reviewer: ```suggestion
   // Create Alice's and Bob's Holding IDs
```

   HoldingIdentity aliceHoldingID = HoldingIdentity.Companion.create(aliceX500);
   HoldingIdentity bobHoldingID = HoldingIdentity.Companion.create(bobX500)
   ```        
5. Use Simulator to create virtual nodes for Alice and Bob from their respective holding identities. The second argument allows you to specify which flows should be loaded up to each virtual node. In this case, Alice runs the initiating flow, with Bob running the responder flow.
   ```java

>    // Create Alice and Bob's virtual nodes, including the Class's of the flows which will be registered on each node.

Reviewer: ```suggestion
   // Create Alice's and Bob's virtual nodes, including the slasses of the flows which will be registered on each node
```

>    // We don't assign Bob's virtual node to a variable because we don't need it for this particular test.

Reviewer: ```suggestion
   // We don't assign Bob's virtual node to a variable because we don't need it for this particular test
```

   ```java
   // Create an instance of the MyFirstFlowStartArgs which contains the request arguments for starting the flow
   MyFirstFlowStartArgs myFirstFlowStartArgs = new MyFirstFlowStartArgs(bobX500);
   ```
   If running this flow on Corda itself, you would send the following `requestBody` over HTTP-RPC:
   ```http
    {
       "clientRequestId": "r1",
       "flowClassName": "com.r3.developers.csdetemplate.MyFirstFlow",
       "requestData": {"otherMember":"CN=Bob, OU=Test Dept, O=R3, L=London, C=GB","message":"Hello Bob"}
   }
   ```
   When using Simulator in tests, use the `RequestData` class, which simulates the `requestBody`:
   ```java
   // Create a requestData object
   RequestData requestData = RequestData.Companion.create(
           "request no 1",        // A unique reference for the instance of the flow request
           MyFirstFlow.class,              // The name of the flow class which is to be started
           myFirstFlowStartArgs            // The object which contains the start arguments of the flow
   );
   ```
7. Pass the `requestData` to the `callflow()` function on Alice’s virtual node, to trigger the flow on Simulator.

>    Simulator simulates running the flow and responder flow, substituting simulated services as required, and returning a flow response String.

Reviewer: ```suggestion
   Simulator simulates running the flow and responder flow, substituting simulated services as required, and returning a flow response string.
```

   ```java

>    // Call the Flow on Alice's virtual node and capture the response from the flow

Reviewer: ```suggestion
   // Call the flow on Alice's virtual node and capture the response from the flow
```

   ```java
   // Create Alice's and Bob HoldingIDs
   HoldingIdentity aliceHoldingID = HoldingIdentity.Companion.create(aliceX500);
   HoldingIdentity bobHoldingID = HoldingIdentity.Companion.create(bobX500)
   ```        
5. Use Simulator to create virtual nodes for Alice and Bob from their respective holding identities. The second argument allows you to specify which flows should be loaded up to each virtual node. In this case, Alice runs the initiating flow, with Bob running the responder flow.
   ```java
   // Create Alice and Bob's virtual nodes, including the Class's of the flows which will be registered on each node.
   // We don't assign Bob's virtual node to a variable because we don't need it for this particular test.
   SimulatedVirtualNode aliceVN = simulator.createVirtualNode(aliceHoldingID, MyFirstFlow.class);
   simulator.createVirtualNode(bobHoldingID, MyFirstFlowResponder.class);
   ```
   You can read more about initiating flows and responder flows in the section on [Your first flow ](../first-flow.html#initiating-and-responding-flows).

> 6. Create the arguments to pass to the flow. In the flow file `MyFirstFlowStartArgs.kt`, create a class `MyFirstFlowArguments` specifically for holding the flow start arguments:

Reviewer 2: "In the flow file `MyFirstFlowStartArgs.kt`"
should be
"In the flow file `MyFirstFlowStartArgs.java`"

I think
"create a class `MyFirstFlowArguments`"
should be
"create a class `MyFirstFlowStartArgs`"

   ```java
   // A class to hold the arguments required to start the flow
   public class MyFirstFlowStartArgs {
     ...
   }
   ```
   <!--no hello bob??-->
   Use the same class to specify that the message needs to go to Bob and the message should be "Hello Bob":
   ```java
   // Create an instance of the MyFirstFlowStartArgs which contains the request arguments for starting the flow

>    MyFirstFlowStartArgs myFirstFlowStartArgs = new MyFirstFlowStartArgs(bobX500);

Author: @ceebee-dev can you confirm this is correct? Should the message be an argument?

Reviewer 2: That's correct.

Author: I think we're saying to use the same class as created in step 6?
So:
"Use the `MyFirstFlowStartArgs` class to specify that the message should go to Bob."
?

Reviewer 2: Yes:
"Use the MyFirstFlowStartArgs class to specify that the message should go to Bob."
or 
"Use the same class to specify that the message needs to go to Bob "
Is what we need
